File: server/pipeline/sessionManager.ts
import type { PipelineStage, PipelineTask, StageRun } from '../../src/types.js'
import { execFileSync } from 'node:child_process'
import process from 'node:process'
import { findStageRunBySessionId, updateStageRun } from '../db/stageRunsRepo.js'
import { IS_LINUX } from '../platform.js'

const ZOMBIE_STATE_RE = /^State:\s+Z/m

/**
 * Build a human-readable session name from task + stage + iteration.
 * Format: {slug}-{stage}-iter-{n}  (e.g. fix-login-bug-umsetzung-iter-3)
 */
export function buildSessionName(task: PipelineTask, stage: PipelineStage, iteration: number): string {
  return `${task.slug}-${stage}-iter-${iteration}`
}

/**
 * Check whether a given PID is still alive (without sending a real signal).
 * Returns false if the process has exited or PID is null.
 *
 * EPERM handling: process.kill(pid, 0) throws EPERM when the PID exists but
 * is owned by a different user. We treat this as "alive" to avoid wrongly
 * restarting a foreign process on top of it. A PID-reuse race (same PID, now
 * a foreign process) would then be misclassified as alive — the recovery
 * path gates further on session_id existence to recover safely.
 */
function isPidZombie(pid: number): boolean {
  try {
    if (IS_LINUX) {
      const stat = execFileSync('cat', [`/proc/${pid}/status`], { encoding: 'utf8', timeout: 500 })
      return ZOMBIE_STATE_RE.test(stat)
    }
    else {
      // macOS: ps -p PID -o stat= returns the state character(s); 'Z' = zombie
      const stat = execFileSync('ps', ['-p', String(pid), '-o', 'stat='], { encoding: 'utf8', timeout: 500 }).trim()
      return stat.startsWith('Z')
    }
  }
  catch {
    return false
  }
}

export function isPidAlive(pid: number | null): boolean {
  if (pid === null || pid <= 0)
    return false
  try {
    process.kill(pid, 0)
    // kill(0) succeeds for zombie processes too — verify the process isn't defunct
    return !isPidZombie(pid)
  }
  catch (err) {
    if ((err as NodeJS.ErrnoException).code === 'EPERM')
      return true
    return false
  }
}

/**
 * Decide whether a running stage_run can be reconnected (live PID) or must
 * be resumed (dead PID but session_id still exists on disk).
 */
export interface RecoveryDecision {
  kind: 'alive' | 'resume' | 'restart'
  reason: string
}

export function decideRecovery(stageRun: StageRun): RecoveryDecision {
  if (isPidAlive(stageRun.pid))
    return { kind: 'alive', reason: `PID ${stageRun.pid} still running` }
  if (stageRun.sessionId)
    return { kind: 'resume', reason: `session ${stageRun.sessionId} available for --resume` }
  return { kind: 'restart', reason: 'no live PID and no session — must start fresh' }
}

/**
 * Persist a session id onto a stage run when it becomes known. Safe no-op
 * if already set to the same value.
 */
export function attachSessionId(stageRunId: string, sessionId: string): void {
  const existing = findStageRunBySessionId(sessionId)
  if (existing) {
    // Already attached (same run = no-op; different run = don't steal it).
    return
  }
  updateStageRun(stageRunId, { sessionId })
}

File: server/internal/pipeline/session_manager.go
package pipeline

import (
	"context"
	"fmt"

	"github.com/lx-wnk/agent-dashboard/server/internal/db/ent"
	"github.com/lx-wnk/agent-dashboard/server/internal/db/repo"
	"github.com/lx-wnk/agent-dashboard/server/internal/proc"
)

type RecoveryDecision struct {
	Kind   string
	Reason string
}

// stageRunPid returns the PID recorded on the stage_run, or 0 when none was.
// proc.IsPidAlive(0) is false, so an absent PID reads as "not alive" — correct
// only where a PID-less run really is a zombie. It is not correct for an
// HTTP-adapter stage, which runs with no local process at all; see the
// process-lifetime guard in sweepOrphanRuns case 4.
func stageRunPid(sr *ent.StageRun) int {
	if sr.Pid == nil {
		return 0
	}
	return *sr.Pid
}

func DecideRecovery(sr *ent.StageRun) RecoveryDecision {
	pid := stageRunPid(sr)
	if proc.IsPidAlive(pid) {
		return RecoveryDecision{Kind: "alive", Reason: fmt.Sprintf("PID %d still running", pid)}
	}
	if sr.SessionID != nil && *sr.SessionID != "" {
		return RecoveryDecision{Kind: "resume", Reason: fmt.Sprintf("session %s available for --resume", *sr.SessionID)}
	}
	return RecoveryDecision{Kind: "restart", Reason: "no live PID and no session — must start fresh"}
}

func BuildSessionName(taskSlug, stage string, iteration int) string {
	return fmt.Sprintf("%s-%s-iter-%d", taskSlug, stage, iteration)
}

func AttachSessionID(ctx context.Context, stageRunID, sessionID string, srRepo repo.StageRunRepo) error {
	existing, err := srRepo.GetBySessionID(ctx, sessionID)
	if err == nil && existing != nil {
		return nil
	}
	sid := sessionID
	_, err = srRepo.Update(ctx, stageRunID, repo.UpdateStageRunInput{SessionID: &sid})
	if err != nil {
		return fmt.Errorf("AttachSessionID: %w", err)
	}
	return nil
}
